build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/transport/TransportVersionManagementPlugin.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.gradle.internal.transport;
+
+import org.elasticsearch.gradle.util.GradleUtils;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.tasks.SourceSet;
+
+public class TransportVersionManagementPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(Project project) {
+        String transportVersionReferencesFile = "transport-version/references.txt";
+        var collectTask = project.getTasks()
+            .register("collectTransportVersionReferences", CollectTransportVersionReferencesTask.class, t -> {
+                t.setGroup("Transport Versions");
+                t.setDescription("Collects all TransportVersion references used throughout the project");
+                SourceSet mainSourceSet = GradleUtils.getJavaSourceSets(project).findByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                t.getClassPath().setFrom(mainSourceSet.getRuntimeClasspath());
```

Comment:
I don't think we want to use the runtime classpath here. This is going to include the modules' code, plus all its dependencies. Since we're using ASM to read the bytecode, and not actually loading classes, we don't actually need any dependencies, we just need the classfiles which we can get via `SourceSet.getOutput()`.

Reply:
We do actually need the dependencies because of how some projects are broken up into sub libs, eg esql has compute/etc, where definitions might exist. We could alternatively apply this plugin to every elasticsearch.build project, but that includes tests etc which would be a bunch of extra tasks/work that could never have these TV references.

Reply:
> We could alternatively apply this plugin to every elasticsearch.build project, but that includes tests etc which would be a bunch of extra tasks/work that could never have these TV references.

I don't think that's an issue. We only care about the `main` sourceset and "test only" projects (i.e. QA projects) don't apply this plugin. We only apply it to "production" projects. I think even if we maybe scan some code projects that aren't applicable, that's _way_ preferrable to the current implementation that ends up scanning all our dependencies many times over.

Reply:
There's actually another problem with this approach, we'll end up with duplicate declarations whenever we depend on some other project that also declares TVs as we don't distinguish between stuff declared in this project or a dependent one since we're just smashing everything onto the runtime classpath.

Reply:
> Do we really need to peak at server and lucene dozens and dozens of times?
we won't. We only look at the _local_ dependencies of the ES plugin. server and lucene are compile only, they won't be on the runtime classpath

Reply:
>  we'll end up with duplicate declarations

There can't be duplicates, since those the definitions exist in server. There can be references to the same definition multiple times, but that is fine. We only collect these for validation.

Reply:
Don't some plugins depend on others in ways that bring them in on the runtime classpath?

Reply:
In either case, I still think it's worth making this more efficient and avoiding scanning classes that aren't local code. I think we can do that relatively easily w/o any major changes here. Basically, if we're pointing to class files or jars outside the of the project workspace, we can safely ignore those. Some plugins bring in a good chunk of dependencies (aws/gcp/etc SDKs) and we don't wanna crack all those open and look for TVs.

Reply:
Alright, I've changed the global plugin to look for the non-global plugin instead of PluginBuildPlugin, and I've explicitly applied that plugin to the example in esql/compute. Most of the time we won't have to explicitly apply the plugin, it should just work as part of ES plugin.